File: Honors/02/honors02.py
# Honors Project 02
# Creates a population of bitstrings and evaluates their fitness
# Algorithm
'''
For 100 individuals
    Generate bitstring
        100 bits
        Random
        Print
    Evaluate fitness
        Count 1s
        Store
'''

# Imports
import datetime, pdb, random, string

# Functions
def fitness(individual):
    '''
    Evaluates fitness based on number of 1 bits.

    Arguments: Bitstring
    Return: Fitness float
    Algorithm:
        Initialize count
        For each bit in string
            If 1, increment count
    '''
    count = 0
    for i in individual:
        if i == '1':
            count += 1
    return count
        

def genIndividual():
    '''
    Creates a 100 bit string of random bits.

    Returns: Individual array
    Algorithm:
        Seed random
        Create bitstring
            For 100 chars
                Randomly append 1 or 0 to bitList
            Merge bitList
        Create individual
            Append bitList
            Assign fitness
    '''
    # Create Bitstring
    random.seed()
    bitList = []
    for i in range(1,101):
        bitList.append(str(random.randint(0,1)))
    bitString = ''.join(bitList)
    # Create individual
    individual = [bitString, 0.0]

    return individual

# 100 individuals
population = []
for i in range(0,101):
    population.append(genIndividual())

# Evaluate fitness
for individual in population:
    individual[1] = fitness(individual[0])

# Print individuals
for individual in population:
    print individual

File: Honors/02/honors02.cpp
// GA Project 02
// Create a population of bitstrings and evaluates fitness

/* Algorithm */
/*
100 individuals
    Array of individual()
For each individual
    Print
*/

/* Includes */
#include <iostream>
#include <stdlib.h>
#include <string>
#include <sys/time.h>

/* Using */
using namespace std;

/* Classes */
class individual{
    public:
        string bitstring;
        int fitness;

        individual();

        friend ostream &operator<<(ostream &stream, individual ob);

        string genBitstring(){
            string temp;
            for(int i = 0; i < 100; i++){
                int rChar = (rand() % 2);
                if (rChar == 0){
                    temp += "0";
                }
                else if (rChar == 1){
                    temp += "1";
                }
            }
            return temp;
        }

        int genFitness(){
            int count = 0;
            for (int i = 0; i < bitstring.length(); i++){
                if (bitstring[i] == '1'){
                    count++;
                }
            }
            return count;
        }
};

individual::individual(){
    bitstring = genBitstring();
    fitness = genFitness();
}

ostream &operator<<(ostream &stream, individual ob){
    stream << "Bitstring: " << ob.bitstring << "\nFitness: " << ob.fitness << "\n\n";
    return stream;
}


/* Main */
int main(){
    //Seed RNG
    timeval t1;// To seed rand()
    gettimeofday(&t1, NULL);// Microsecond resolution
    srand(t1.tv_usec * t1.tv_sec);
    individual population [100];
    for (int i = 0; i < 100; i++){
        population[i] = individual();
    }
    for (int i = 0; i < 100; i++){
        cout << population[i] << endl;
    }

    return 0;
}
